File: auth/jwt_create.R
#!/usr/bin/env Rscript

#' Firstbeat Sports Cloud API - JWT Generation Script
#'
#' Usage:
#'   Rscript jwt_create.R <consumer_id> <shared_secret>
#'
#' Prerequisites:
#'   install.packages("openssl")
#'   install.packages("jose")

library(openssl)
library(jose)

createJwtToken <- function(consumerId, sharedSecret) {
  tryCatch({
    now <- as.numeric(Sys.time())
    after_five_minutes <- now + 300  # 5 minutes validity

    claim <- jwt_claim(iss = consumerId, iat = now, exp = after_five_minutes)
    secret <- charToRaw(sharedSecret)

    token <- jwt_encode_hmac(claim, secret)
    return(token)
  }, error = function(e) {
    cat("Error generating token:", conditionMessage(e), "\n", file = stderr())
    quit(status = 1)
  })
}

main <- function() {
  args <- commandArgs(trailingOnly = TRUE)

  if (length(args) != 2) {
    cat("Usage: Rscript jwt_create.R <consumer_id> <shared_secret>\n", file = stderr())
    quit(status = 1)
  }

  consumerId <- args[1]
  sharedSecret <- args[2]

  token <- createJwtToken(consumerId, sharedSecret)
  cat("Bearer", token, "\n")
}

main()

File: auth/JwtCreate.java
/**
 * Firstbeat Sports Cloud API - JWT Generation Example
 * 
 * Dependencies (Maven coordinates):
 *   - com.auth0:java-jwt:4.4.0 (or newer)
 * 
 * Compile & Run:
 *   javac -cp "path/to/java-jwt.jar" JwtCreate.java
 *   java -cp ".:path/to/java-jwt.jar" JwtCreate <consumer_id> <shared_secret>
 */

import java.util.Calendar;
import java.util.Date;
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.exceptions.JWTCreationException;

public class JwtCreate {

    public static String createJwtToken(String consumerId, String sharedSecret) {
        try {
            // Algorithm is HMAC256
            Algorithm algorithm = Algorithm.HMAC256(sharedSecret);
            
            Calendar c = Calendar.getInstance();
            Date now = c.getTime();
            
            // Expires in 5 minutes (300 seconds)
            c.add(Calendar.SECOND, 300);
            Date expires = c.getTime();

            return JWT.create()
                .withIssuer(consumerId)
                .withIssuedAt(now)
                .withExpiresAt(expires)
                .sign(algorithm);
                
        } catch (IllegalArgumentException e) {
            System.err.println("Error: Shared secret cannot be null or empty.");
            return null;
        } catch (JWTCreationException e) {
            System.err.println("Error creating JWT: " + e.getMessage());
            return null;
        }
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            System.out.println("Usage: java JwtCreate <consumer_id> <shared_secret>");
            System.exit(1);
        }

        String consumerId = args[0];
        String sharedSecret = args[1];

        String token = createJwtToken(consumerId, sharedSecret);
        
        if (token != null) {
            System.out.println("Bearer " + token);
        } else {
            System.exit(1);
        }
    }
}
